server/controllers/events/events_controller_test.go

```diff
@@ -618,6 +618,73 @@ func TestPost_AzureDevopsPullRequestDeletedCommentIgnoreEvent(t *testing.T) {
 	})
 }
 
+func TestPost_AzureDevopsPullRequestCommentPassingIgnores(t *testing.T) {
+	u := "user"
+	user := []byte(u)
+
+	t.Log("when the event should not be ignored it should succeed until parsing section")
```

Comment:
Not sure i understand this. Why should the code succeed until the parsing section? It's probably a naive question but shouldn't the code always succeed? Perhaps it's this test description that's confusing to me

Reply:
The process is that after the ignore tests, it will go to the Repo URL provided in the web request and get all the data about the PR directly from the ADO API. 
For this test we define the web request's json and pass it into the test. To have the test pass completely then we would need to pass it a repo url / pull request which is completely open / accessible to any machine that runs the test. Definitely possible if we make an ADO Public Repo but that may require more thought than just me making one and pointing it. 

Because of this, the test is limited as it can only pass until it gets to the point where it talks to the API. After that it has no choice but to fail unless we can give it a real test URL to talk to.

Reply:
Instead of supplying it a real test URL, isn't that when we would mock the API call?

Reply:
Do you have an example? Happy to try that instead - I was just going off the ADO tests that were already there

Reply:
We use pegomock in the code base. It's used in many of our tests but perhaps none were added to ADO. Please look at the other examples.

As far as i know, pegomock is used to create the mocks, and then you can use import to import the mocks, and finally the function calls to 3rd parties can have mocked returns.

https://github.com/runatlantis/atlantis/blob/2d29db68932ea9f128c19ff6fc84dda0bb86ff1e/Makefile#L26

https://github.com/runatlantis/atlantis/blob/2d29db68932ea9f128c19ff6fc84dda0bb86ff1e/server/events/vcs/client.go#L20

https://github.com/runatlantis/atlantis/tree/2d29db68932ea9f128c19ff6fc84dda0bb86ff1e/server/events/vcs/mocks

https://github.com/runatlantis/atlantis/blob/2d29db68932ea9f128c19ff6fc84dda0bb86ff1e/server/events/vcs/mocks/mock_github_credentials.go

https://github.com/runatlantis/atlantis/blob/2d29db68932ea9f128c19ff6fc84dda0bb86ff1e/server/jobs/job_url_setter_test.go#L12

https://github.com/runatlantis/atlantis/blob/2d29db68932ea9f128c19ff6fc84dda0bb86ff1e/server/jobs/job_url_setter_test.go#L27

Reply:
Alrighty, I'll start to have a look through this. 
Feel free to revert the initial change if you need to quickly, depending on how complex this is it might take me a little while to write a better test.

Reply:
We really do not want the code to throw an error and would like to mock the api whenever necessary.

We're hoping to release early next week so if you think you can figure it out before then it would really improve this pr.

cc @jamengual @krrrr38 @jsw1993  (correct me if im wrong)

Reply:
Thank you again @SSKLCP. Your contributions imprpve the code base a lot more than you may realize. :smile:

Reply:
So I realise that this test is using pegomock to mock a request to the Atlantis API. But when processing that call to the Atlantis API it then goes off to the ADO API to get more info about the PR.
Not sure if there are such things as nested mocks that you've used elsewhere? 
(I could be missing the point slightly - This is getting a little bit out of my realm of knowledge) 

But I understand the desire to not have a passing failing test, it does feel a bit bad.

Reply:
This PR goal is to fix the panic when commenting on azure PR. The current test case does not use `IsDeleted` logic, because if parse failed, [does not execute e.HandleAzureDevopsPullRequestCommentedEvent](https://github.com/runatlantis/atlantis/blob/main/server/controllers/events/events_controller.go#L269-L277
). So this test cannot test the original issue. 

I guess `payload.PullRequest` is invalid. The following json is correct.

```json
		"resource": {
			"comment": {...},
			"pullRequest": {
				"pullRequestId": 1,
				"repository": {...}
			}
```

<details>
<summary>After that, you can test `e.HandleAzureDevopsPullRequestCommentedEven` like followings</summary>

`e.Parser` is already tested on `event_parser_test.go, so just mocked.

```diff
diff --git a/server/controllers/events/events_controller_test.go b/server/controllers/events/events_controller_test.go
index ca67f366..d0aa7284 100644
--- a/server/controllers/events/events_controller_test.go
+++ b/server/controllers/events/events_controller_test.go
@@ -651,6 +651,9 @@ func TestPost_AzureDevopsPullRequestCommentPassingIgnores(t *testing.T) {
 		VCSClient:                       vcsmock,
 	}
 
+	repo := models.Repo{}
+	When(e.Parser.ParseAzureDevopsRepo(matchers.AnyPtrToAzuredevopsGitRepository())).ThenReturn(repo, nil)
+
 	payload := `{
 		"subscriptionId": "11111111-1111-1111-1111-111111111111",
 		"notificationId": 1,
@@ -666,9 +669,11 @@ func TestPost_AzureDevopsPullRequestCommentPassingIgnores(t *testing.T) {
 				"commentType": "text",
 				"content": "test"
 			},
-			"PullRequest": {
+			"pullRequest": {
 				"pullRequestId": 1,
-				"Repository": ""
+				"repository": {
+					"id": "1"
+                }
 			}
 		}
 	}`
@@ -678,10 +683,9 @@ func TestPost_AzureDevopsPullRequestCommentPassingIgnores(t *testing.T) {
 		req.Header.Set(azuredevopsHeader, "reqID")
 		When(v.Validate(req, user, secret)).ThenReturn([]byte(payload), nil)
 		w := httptest.NewRecorder()
-		e.Parser = &events.EventParser{}
 		e.Post(w, req)
 		t.Log(w)
-		ResponseContains(t, w, http.StatusBadRequest, "Failed parsing webhook: json: cannot unmarshal string into Go struct field")
+		ResponseContains(t, w, http.StatusOK, "")
 	})
 }
 ```

</details>

If apply adove diff, you can check `IsDeleted` works correctly.

Reply:
Thank you so much for this @krrrr38! 
I've made the changes here 544c9468cf6cf85ea5a57b520a3aaa03556ccb63 and it's now passing with a 200 :)

I don't suppose I can ask for you to explain the effect that removing `e.Parser = &events.EventParser{}` had on this? Was the parser mocked as it was defined in `e := events_controllers.VCSEventsController{...}` and then specifying it below was asking it to use the real Event Parser instead? 

Definitely seems pretty powerful, I'll try to use it in my other tests. 

Thank you again 😄

Reply:
> I don't suppose I can ask for you to explain the effect that removing e.Parser = &events.EventParser{} had on this? Was the parser mocked as it was defined in e := events_controllers.VCSEventsController{...} and then specifying it below was asking it to use the real Event Parser instead?

I'm not sure the original impl, but it seems weird to me to pass mock into `events_controllers.VCSEventsController{}` and override it. Generally this is a unit test for controller, so parser can be mocked if we test parser in the parser unit test.